[wgsl] Add workgroup_size validation tests.

dj2 opened the pull request:
This CL adds validation tests for the `workgroup_size` attribute.

Fixes #1691

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

zoddicus commented:
Another corner, since floats are allowed, what happens if a fractional value, like 2.5 is supplied?

dj2 commented:
Added test for mixing `u` and `i`. Added test for the max value listed in the spec `(256, 256, 64)`. Added `0` and negative value tests.

Did not test really large values as that's a pipeline-creation error and an adapter value. The float is not a valid value, which is tested already that floats are rejected.

This is testing the `0` and negative values here, which means they're shader-creation errors. Per spec, they're listed as pipeline-creation errors. Is it valid to error at shader-creation for them or does it have to be pipeline-creation (@dneto0)

zoddicus reviewed (commented):
lgtm, but I will let @dneto0 do the final approval

Sorry for misreading the float tests.

zoddicus reviewed (commented):
Trying the clear the changes requested bit

zoddicus reviewed (approved):
Apparently I have to approve to clear the changes requested bit.

alan-baker commented on src/webgpu/shader/validation/shader_io/workgroup_size.spec.ts:
```diff
@@ -0,0 +1,153 @@
+    pass: false,
+  },
+
+  max_values: {
+    src: `@workgroup_size(256, 256, 64)`,
+    pass: true,
+  },
```
This is still reasonably likely to fail on many implementations due to the number of invocations limit. So maybe we should have a test that validates beyond these numbers?

dj2 replied:
These numbers are the values listed in the WebGPU spec, so I believe they are the most that needs to be supported. These are all validation tests so they won't be executed on actual devices, but the values should compile I believe.

alan-baker replied:
maxComputeInvocationsPerWorkgroup is 256 for reference. So (256,256,64) is too large. Not a big deal since it's only a validation test, but likewise we could ensure larger values are allowed (someone could also ask for an adaptor with higher limits).

dj2 replied:
I wasn't sure if requesting more would be portable, because the adapter can reject the higher limit, no? So, we can only validate that the minimum passes, otherwise it's out of our hands?

Or, since that's validation outside WGSL, we can just treat this as an i32 and pass in whatever and leave it for execution tests to catch big values?

alan-baker replied:
I expect https://github.com/gpuweb/gpuweb/pull/4073 to be approved. So there is no upper bound to hit for a shader-creation error. So unless you're checking for pipeline errors the test could (but don't have to) cover larger values. Not really asking for a change here just more as an FYI. The main point of the review was the request for expression (not just declaration or literal) based values.

alan-baker reviewed (commented):
Suggest adding some expression variants for both const-expression and override-expression.

dj2 commented:
Added `const_expr` and `override_expr` tests. PTAL.
